PolyEngine/Core/Src/BTreeMap.hpp

```diff
@@ -0,0 +1,532 @@
+#pragma once
+
+#include "BTreeNode.hpp"
+#include "Optional.hpp"
+
+namespace Poly {
+
+	template<typename K, typename V, size_t B>
+	struct Range { //note(vuko): doubles as a proper fucking iterator; behold the goddamn beauty! (alright, this is for show only for now)
+		using BTree  = Impl::BTree<K, V, B>;
+		using KVERef = typename BTree::KVERef;
+
+		KVERef front;
+		KVERef back;
+
+		struct KV {
+			const K& key;
+			V& value;
+		};
+
+		KV next_unchecked() {
+			KVERef kv_ref = this->front;
+
+			if (kv_ref.idx < kv_ref.node_ref.node->len) {
+				front = KVERef{kv_ref.node_ref, kv_ref.idx + 1};
+				return {kv_ref.node_ref.node->keys[kv_ref.idx], kv_ref.node_ref.node->values[kv_ref.idx]};
+			}
+
+			do {
+				kv_ref = kv_ref.node_ref.ascend().parent;
+			} while (kv_ref.idx >= kv_ref.node_ref.node->len);
+
+			front = BTree::first_leaf_edge(KVERef{kv_ref.node_ref, kv_ref.idx + 1}.descend());
+			return {kv_ref.node_ref.node->keys[kv_ref.idx], kv_ref.node_ref.node->values[kv_ref.idx]};
+		}
+
+		KV prev_unchecked() {
+			KVERef kv_ref = this->back;
+
+			if (kv_ref.idx > 0) {
+				this->back = KVERef{kv_ref.node_ref, kv_ref.idx - 1};
+				return {kv_ref.node_ref.node->keys[kv_ref.idx], kv_ref.node_ref.node->values[kv_ref.idx]};
+			}
+
+			do {
+				kv_ref = kv_ref.node_ref.ascend().parent;
+			} while (kv_ref.idx == 0);
+
+			this->back = BTree::last_leaf_edge(KVERef{kv_ref.node_ref, kv_ref.idx - 1}.descend());
+			return {kv_ref.node_ref.node->keys[kv_ref.idx], kv_ref.node_ref.node->values[kv_ref.idx]};
+		}
+
+		KV next() {
+			assert(!(front == back));
+			return next_unchecked();
+		}
+	};
+
+	//todo(vuko): docs
+	template<typename K, typename V, size_t Bfactor = 6>
+	class BTreeMap {
```

Comment:
Eh, `Map` is a bit too undescriptive. I mean, is a `HashMap` not a `Map`? ;) Then there's the issue of how they're different. I was initially thinking in the lines of `HashMap` and `TreeMap`, 
and since `Tree` is literally one character away from `BTree`, here we are. It's good to be clear about things. Well, I could settle for `SortedMap`/`UnorderedMap`, for example.